### ISSUE TITLE
New block: Accordion block

### ISSUE THREAD

**Issue by mapk** [WordPress/gutenberg#21584]
There's been a few requests for an Accordion block in core. Let's start exploring how this looks.

**Requirements**
- Pure CSS.
- Add accordion headers with arrow icons.
- Allow any block to be added within an accordion section.

Is the accordion section simply a Group block that is able to be collapsed? Basically it would work similarly to the Settings sidebar sections.

![accordion](https://user-images.githubusercontent.com/617986/79265577-2cf32280-7e4b-11ea-9a1b-d39ace25d40e.gif)

Kudos to @melchoyce for this idea. Some examples she shared recently:

https://codepen.io/raubaca/pen/PZzpVe
https://codepen.io/vinsongrant/pen/qbGKed
https://codepen.io/abergin/pen/ihlDf

**Comment by paaljoachim** [comment]
Great! Yeah this is needed. I have seen a few requests over a longer period of time asking for an accordion block.

**Comment by carolinan** [comment]
If there is a fallback for IE, it could be a details element, reducing the markup drastically.
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details

**Comment by aristath** [comment]
Actually accordions don't need CSS, JS etc... We could simply use a `<details>` HTML element.
```html
<details>
  <summary>This is the accordion title</summary>
  This is the accordion content
</details>
```
^ The above code would render like this:

<details>
  <summary>This is the accordion title</summary>
  This is the accordion content
</details>

Coincidentally I started building a `details` block yesterday and it works fine that way.

We _can_ add options to change the styling of the triangle etc if needed, and there's some extra info on [this article](https://css-tricks.com/quick-reminder-that-details-summary-is-the-easiest-way-ever-to-make-an-accordion/)

**Comment by mtias** [comment]
We could do a render fallback on the server for IE, or also consider not doing anything (it'll still render the content, just expanded by default).

**Comment by aristath** [comment]
having it expanded by default in IE is not a big issue... As long as it renders and looks OK I don't think we should add any extra scripts etc to make IE behave like modern browsers. Supporting IE is not the same as making it look & behave identically :+1:

**Comment by paaljoachim** [comment]
Here are some plugin examples that use accordion blocks.
Twenty Twenty. 
Gutenberg plugin activated.

**Atomic Blocks**
https://wordpress.org/plugins/atomic-blocks/

Backend:
<img width="1001" alt="Atomic-Blocks-Accordion" src="https://user-images.githubusercontent.com/5323259/83339922-b0b97f00-a2d2-11ea-9874-9a4913033c4c.png">

Title selected:
<img width="870" alt="Atomic-Blocks-Accordion-title-selected" src="https://user-images.githubusercontent.com/5323259/83340248-11968680-a2d6-11ea-8372-2a299a67c674.png">

Content selected:
<img width="871" alt="Atomic-Blocks-Accordion-content-selected" src="https://user-images.githubusercontent.com/5323259/83340258-312daf00-a2d6-11ea-85d5-d96b02329f5d.png">


Frontend:
<img width="1014" alt="Atomic-Blocks-Accordion-frontend" src="https://user-images.githubusercontent.com/5323259/83339949-17d73380-a2d3-11ea-8ad7-866e7c861857.png">

-----------

**Co Blocks**
https://wordpress.org/plugins/coblocks/

Backend.

Title is selected.
<img width="919" alt="Co-Blocks-Accordion-1" src="https://user-images.githubusercontent.com/5323259/83340059-1c501c00-a2d4-11ea-876b-8cd31a57cc06.png">

Content is selected:
<img width="921" alt="Co-Blocks-Accordion-2" src="https://user-images.githubusercontent.com/5323259/83340068-2a05a180-a2d4-11ea-95da-48aa9d9d5ddc.png">

To add a new accordion one clicks the Inserter area similar to the Group Block.


Frontend:
<img width="729" alt="Co-Blocks-Accordion-frontend" src="https://user-images.githubusercontent.com/5323259/83340098-849efd80-a2d4-11ea-9dd4-688bbe79b4f8.png">

-----------

**Accordion Blocks by Phil Buchanan**
https://wordpress.org/plugins/accordion-blocks/

Backend.

Title selected.
<img width="880" alt="Accordion-Block-Phil-1" src="https://user-images.githubusercontent.com/5323259/83340163-2c1c3000-a2d5-11ea-9933-d6331b88966a.png">

Content selected.
<img width="879" alt="Accordion-Block-Phil-2" src="https://user-images.githubusercontent.com/5323259/83340174-48b86800-a2d5-11ea-876a-9b4dbd007870.png">

Frontend.
There is a minus on the right side when the accordion is open. When the accordion is closed there is a + sign.
<img width="706" alt="Accordion-Block-Phil-frontend" src="https://user-images.githubusercontent.com/5323259/83340206-a8167800-a2d5-11ea-82d3-b085667008d2.png">

**Comment by ZebulanStanphill** [comment]
It's worth noting that using heading elements in the `<summary>` is valid, and that would be the correct way to handle a lot of use-cases for accordions like step-by-step how-to articles.

Should the Accordion block have an option to switch the contents of the `<summary>` from plain (no inner elements other than text) to different heading levels (similar to the options provided in the Heading block); or should the `<summary>` be an entire `InnerBlocks` area where you can insert whatever typography-related block you want?

I'm currently leaning towards the former solution, since it's less complicated, and it's much easier to migrate from that to the latter solution in the future than to go the other way around.

**Comment by LukaszJaro** [comment]
The only thing with IE11 and `<details>` HTML element is they are expanded by default as mentioned. The problem with that is it defeats the purpose of toggles.

**Comment by aristath** [comment]
IE11 doesn't _break_ with a `<details>` element... and sure, it's not the same as it is in a modern browser but that's not a dealbreaker. 
Nobody on IE expects to see the same thing they'd see when using Firefox/Chrome/Brave/Whatever. Different browsers exists for different people, different tastes and needs or limitations. If they use IE, then the rest of the planet shouldn't suffer for it. 
A workaround could be implemented for IE by loading a conditional script that would tweak the behaviour, but... is it really worth it? Adding extra html & JS in other browsers just to make IE work would do a lot more damage overall, and we can't really move forward if we keep looking at a broken thing from 7 years ago.
They can still see the content, they can read everything, things are properly styled, just not collapsed by default.

**Comment by LukaszJaro** [comment]
I understand what your saying @aristath I just have a very strange requirements with a project right now which only supports IE11 ! yea...it is an internal site like a intranet type. The content editor wanted some toggles..

And there is still years of support for IE11 :(

https://death-to-ie11.com/

**Comment by enriquesanchez** [comment]
This issue and related PR [were discussed today by the Accessibility team](https://wordpress.slack.com/archives/C02RP4X03/p1595001310427400).

The team suggests to not use the `<details>` element for this scenario, and instead [follow the ARIA Authoring Practices implementation for an accordion component](https://www.w3.org/TR/wai-aria-practices/#accordion)

For more details on why `<details>` is not a suitable element, please see:

1. https://daverupert.com/2019/12/why-details-is-not-an-accordion/
2. https://adrianroselli.com/2019/04/details-summary-are-not-insert-control-here.html

As always, please feel free to ping the team in the #accessibility Make WordPress Slack channel if there are further questions or comments.

**Comment by mapk** [comment]
Here's an iteration on the Accordion block.

![Accordion](https://user-images.githubusercontent.com/617986/87821063-0ff03600-c824-11ea-82fc-fab6a418b354.png)

**Comment by ZebulanStanphill** [comment]
That looks pretty good. One thing we'll need to make sure to include in the block's toolbar is a way to switch the HTML element of the heading, including `h1`-`h6` options, plus maybe even `p`? We should be able to extract the heading level control from the Heading block and reuse it here.

**Comment by mapk** [comment]
Here's a prototype of how I imagine this block working. 

![accordion](https://user-images.githubusercontent.com/617986/88727624-d2718f80-d0e4-11ea-9d81-24f66db88fa4.gif)


- It uses similar block interactions as the Group and Heading blocks.
    - The Accordion heading allows a heading level change.
    - The large `+` Inserter button below works like the Group block.
- The heading is editable on focus of the Accordion block.
- Adding content to the collapsable region requires clicking the `+` Inserter to add a block.
- When resting, the Accordion block collapses to hide the content region.
- When focused, the Accordion block open to show the content.

**Comment by mapk** [comment]
Some recent design feedback lead me to add a couple Style Variations.

### Default

![default](https://user-images.githubusercontent.com/617986/88829152-10bc8c80-d181-11ea-9667-6709cf64c227.png)

### Style Variation 1

![variation 1](https://user-images.githubusercontent.com/617986/88829174-187c3100-d181-11ea-9915-b32ffe1d530f.png)

### Style Variation 2

![variation 2](https://user-images.githubusercontent.com/617986/88829189-1e721200-d181-11ea-8b09-3e66a50466f4.png)

**Comment by mapk** [comment]
Further design feedback from @melchoyce suggested that we just drop a Paragraph block into the content area rather than a big `+` button. The flow would look something like this.

![accordion-with-p](https://user-images.githubusercontent.com/617986/88832609-ecaf7a00-d185-11ea-9ca8-8308f3838e46.gif)

**Comment by mapk** [comment]
Recent design feedback:

### Arrows
The arrows need to be oriented differently. Rather than up/down, they should be right/down.

![lf-arrows](https://user-images.githubusercontent.com/617986/89058903-07fcbf80-d315-11ea-872d-535cc25220ba.png)

### Style variations
There's some question as to whether or not the variations above make sense to include as style variations. I'm waffling back and forth on this one. They are "style" variations that only manipulate styling... so that makes sense to me.

### Default variation
Should the above Style Variation 2 actually be the default because there's basically no styling? It might make it easier to themes to apply their own styles with that as the default.

**Comment by ZebulanStanphill** [comment]
Yeah, I'd favor a default style that's easy for themes to override.

**Comment by paaljoachim** [comment]
Here is a PR for something called the Details Block: 
https://github.com/WordPress/gutenberg/pull/23940
There is some crossover of the Accordion and the Details Blocks.

**Comment by kellychoffman** [comment]
Looking good. Liking the iterations so far and how this is evolving. Some feedback: 

- Add the interaction/flow for adding an additional accordion section
- Add the interaction/flow for adding an additional block within the accordion content
- Add block settings for accordion
To summarize above: I think it would be good to see the full block adding flow in context with the sidebar and its settings

- I would also push the styles a bit more. Right now they are very subtle and I wouldn't consider them a different style, more just removing/adding background+border colors. Try playing more with layout, spacing, position of elements. Similar to this approach: https://github.com/WordPress/gutenberg/issues/22071#issuecomment-624637501

**Comment by pablohoneyhoney** [comment]
+1 to @kellychoffman notes. It's relevant to find the right customisation while adjusting to any theme.
The icon seems a bit too abstract, worth exploring some options there. 

Note that this https://github.com/WordPress/gutenberg/pull/23971 seems to be happening which affects slightly your mock ups.

**Comment by ZebulanStanphill** [comment]
@pablohoneyhoney Not sure if you're referring to the block icon or the expand/collapse icon, but in the case of the latter, I think it's definitely worth trying to find a new icon to indicate expand/collapse. Right now we're using the same icons that we use for the block movers, which is confusing as the two are completely unrelated.

**Comment by pablohoneyhoney** [comment]
Now that you are saying it, both actually @ZebulanStanphill 

I was referring to the block icon. Regarding the collapsable UI (arrow), unsure if there are many other patterns besides arrow or +/-.

**Comment by mapk** [comment]
I've iterated on the UX flow for this block a bit more to help explain how a new block would be added, and how to add another Accordion block below. I see each block being its own accordion to keep things simple. Wrapping multiple accordions into a wrapper can get complex in a number of ways, and I'd like to avoid that.

Taking in the recent feedback, I chose the less stylized variation as the base for this block. This does two things:
1. Makes the default state easier for themes to style.
2. Gives me a more difficult beginning to really tune the UX flow.

**UX flow for adding multiple blocks and then adding another Accordion block**

![accordion-flow](https://user-images.githubusercontent.com/617986/89588503-9e8b1e00-d7f8-11ea-974c-75720d1577d3.gif)

The Accordion block in this flow works much like adding multiple blocks in a Column. While the Columns block surfaces a large `+` button, the Accordion block does not, but rather presents the typical Paragraph block's placeholder. So while editing one block, hitting `Enter` will insert a new block below – just like in the Editor.

When clicking the parent (Accordion) block, it shows the selected border that contains both blocks. Very much like how both the Group block and Columns block work.

Deselecting the Accordion block causes it to collapse. In the gif, this is what happens when I have the Accordion block selected, but then hit the `Enter` key. This action adds a new cursor below the Accordion block causing it to collapse. I then type `/accordion` to insert a new Accordion block again.

----

### Next steps 
1. Work on sidebar settings to see which features the Accordion block would be good to have.
    - ie. Color for heading backgrounds
2. Iterate on a few simple style variations that could include borders, different arrow icons, etc. 
3. Iterate on a new block icon.

Both of these will keep the block simple enough to allow themes to easily style it, but also provide necessary block settings that improve customizations. We should keep in mind that customizations with the use of "Design Tools" are different than Style variations that just provide quick and simple variations for a user to then customize further.

**Comment by jasmussen** [comment]
I left some comments related to an accordion block on [this PR around a "Details" block](https://github.com/WordPress/gutenberg/pull/23940#issuecomment-669750634). The gist is I could see a place for both coexisting, mostly because the details block is so minimal while still being useful.

**Comment by enriquesanchez** [comment]
>I could see a place for both coexisting

@jasmussen From a user perspective, whare are the differences in functionality that these two blocks will provide?

**Comment by mariohamann** [comment]
It's great that considerations exist to bring accordions into the Gutenberg-Core! For a project I also created an accordion block.The thing doesn't run perfectly and smoothly yet, but it does its job.  :)

I have made the following experiences and would like to pass them on:
## Advantages of two blocks (accordion and accordion_element)
I decided early on to work with nested blocks, namely with a parent block (=accordion) and a child block (=accordion_element). The advantages are obvious:
1. the option can be offered to have other accordion elements in the frontend close when a single one opens
2. it is also possible to set the heading level for all elements at the same time, which makes accordions much more maintainable. (Use Case: Someone adds a heading above the accordion, whereupon the heading levels of each accordion element must be reduced by one).
3. from my perspective, accordions are often considered as a whole element, e.g. for FAQs or the like It would therefore also be very useful if they existed as an entire block, e.g. when moving elements. The accordion should always be moveable as a whole without workarounds and no other elements should be able to slide between the accordion elements. Concrete example: It seems strange to me if a picture under an accordion would end up inside the accordion by moving it with the arrows at once.
![moving-image](https://user-images.githubusercontent.com/26542182/89705000-a9a98f80-d959-11ea-97a3-9a219860cc69.gif)
4. further accordion_elements can be added so quickly. In our design, we have adjusted the background of the add_button to the color of the button, so that the full width is filled in, giving a stronger resemblance to the accordion_element.
![add-element](https://user-images.githubusercontent.com/26542182/89705006-b1693400-d959-11ea-965a-185f2822e405.gif)


## Icon on the left side?
I don't know how useful it is to position the unfolding icon on the left side by default. For example, if someone wants to include icons for orientation in their title, this will be difficult. In another project I worked with complementary icons to better structure the content.
![icons-right](https://user-images.githubusercontent.com/26542182/89705007-b7f7ab80-d959-11ea-8d23-32ac5f2efc01.gif)

## Moving content in editor
During my development I had to learn that the content area of an accordion_element can be very painful to edit. In your design @mapk it looks like the accordion_element in the editor is unfolded when it is selected. This was also the case in my projects at first. It is a common use case that content is written first, but only then is it decided to put it into an accordion. This makes it necessary to access the InnerBlock area of an accordion_element without the accordion_element being currently selected. I don't think my solution is ideal, but it was the most feasible one: Everything is always unfolded. This way, contents of accordion_element can be moved to another accordion_element.
![moving-elements-1](https://user-images.githubusercontent.com/26542182/89705015-ce9e0280-d959-11ea-8c0a-d67d1af7665a.gif)

## Separation of the title from the content in frontend and backend
Practical experience has shown that it is very useful to clearly distinguish the clickable head area from the rest when activated. This is especially important if, for example, bold print or even headings are used within the content area. The latter would also be absolutely legitimate from a hierarchical perspective, but can lead to parts of the content being larger than the accordion title. To achieve this, it was essential that an accordion_element be linked by a common frame, but at the same time the title and content had to be separated. @mapk: I therefore consider your older designs with the minimal border to be a much better default state and should still be minimal enough to make them customizable by themes. From a usability perspective, it will also only then become clear that the complete title section is clickable - which will hopefully be the case? :)
![headings-in-content](https://user-images.githubusercontent.com/26542182/89705022-e37a9600-d959-11ea-8894-e4f9286b2dc8.gif)

## Appendix
A few more short thoughts I want to share:
- Concerning the icon, I decided on a plus ("Here you can see more"), which turns to an X ("Here you can close") after opening.
- Hover and focus styles are very important here to make it clear to frontend users that interaction is possible here. We have worked with backgrounds, even though this is not quite ideal from an A11y perspective.

I am looking forward to seeing more designs and can't wait to see the result in the core! Thanks for your efforts!

**Comment by mapk** [comment]
I've iterated on some block settings that will benefit the Accordion block. 

- Style variations (minimal or borders)
- Open by default
- Animate opening and closing
- Heading color settings (text and background)

With these settings, and the settings inherent in each block (ie. Heading levels, Paragraph settings, etc.) I think users can customize these fairly well to their liking. 

![accordion-settings](https://user-images.githubusercontent.com/617986/90201664-5f2c7680-dd90-11ea-9169-f2f9fd8633d8.gif)

**Notes**
- As noted earlier, I've updated the Accordion arrow.

----

### Next steps
1. ~Work on sidebar settings to see which features the Accordion block would be good to have.~
    - ~ie. Color for heading backgrounds~
    - ~Toggle to keep accordion open~
2. ~Iterate on a few simple style variations that could include borders, different arrow icons, etc.~
3. Iterate on a new block icon.

**Comment by mariohamann** [comment]
This is great! I like the idea to use color variations for the header! (Seeing this, I would like to be able to define "global settings" on block level so that I don't have to edit all blocks of the same type after a color change. Sort of like global blocks, only at the settings level instead of the content level, i.e. changes to the settings in one block are automatically transferred to others with the same "Global Settings". I will open a new issue. :))

Personally, I still tend to place the arrow on the right side. A quick look in Dribbble and Google Images confirmed my feeling that about 80% place the accordion icon on the right side. At the same time I understand that in this case in "minimal"-style the reference to the text would be a bit too small and the feeling of an accordion could be overlooked. What do you think about the position, @mapk? Maybe it would help to make "border"-style default?

**Comment by kellychoffman** [comment]
Settings feel like they are in a good place now. Nice!

> Add the interaction/flow for adding an additional accordion section
> Add the interaction/flow for adding an additional block within the accordion content

I would just add these two and I think we'll be close. Two more notes:

- Styles: I still feel like just adding a border isn't unique enough. One idea is to explore different layouts and ways to present this type of information. Another is size. Coblock's Shape Divider block is a fun examples to use that themes could then build on top of.
- Subjective / aesthetic / details: There's something about the arrow treatment that feels dated. More old school wp-admin than G2, if you will. Might be the size. Also, the position of the arrow when you add a border or background color to the header is too tight to the left.

**Comment by ZebulanStanphill** [comment]
It is worth noting that if the icon is on the right, it may be more difficult for the user to associate it with the title unless there is also a border around accordion title and icon.

**Comment by mapk** [comment]
This style variation incorporates the block's additional settings that allow it to visually differentiate from the default variation.

### Style variation
- Larger heading text
- Gradient background in heading
- Custom color heading text
- Icon is right justified
- Borders with custom color

**Block variation**

![variation 1 closed](https://user-images.githubusercontent.com/617986/91501921-9ebc8d80-e87b-11ea-8279-4442361c79ab.png)


### Settings sidebar
- Styles
    - Minimal
    - Bold
- Accordion settings
    - [toggle] Open by default
    - [toggle] Animate opening
- Heading color settings
    - Text color
    - Background color
- Icon settings
    - [toggle] Move icon to other side
    - Select from a few icon choices



**Inspector**

![Block+Document Inspector](https://user-images.githubusercontent.com/617986/91503291-f01a4c00-e87e-11ea-9d85-05d274d3224f.png)

**Comment by mapk** [comment]
I'm exploring a way to work in the icon settings into the block's toolbar.

**Comment by mariohamann** [comment]
Very nice! I wanted to add the idea of exactly those kinds of styles ("minimal" with icon on left and "extended" with icon on right and with borders) as well and am happy to see it here now. :) Thanks, @mapk !

**Comment by ZebulanStanphill** [comment]
I think some of these options ("Move icon to other side") are the kind of thing that should only be exposed at a global level, and never on a per-block basis. Allowing users to have accordions across their site using different values for options like those is a recipe for inconsistent design.

**Comment by mtias** [comment]
Agreed, that's been one of the most problematic parts of style variations, even with the "set by default". I think we need to be more meticulous with style variations until they can be represented fully through attributes, so that they can also be updated by the user / theme globally. We also need to be committed to maintaining the styles, which has created some issues with things like pullquotes.

**Comment by kellychoffman** [comment]
This was sort of what I was getting at with my comments on pushing the style options: to have the styles be a way to present an alternative layout (among other things). The "bold" style is getting there, but the gradient and color used have it appearing a bit too dated.

**Comment by mariohamann** [comment]
> The "bold" style is getting there, but the gradient and color used have it appearing a bit too dated.

+1

Although I proposed to set some background for the heading area on hover or focus, I have the feeling there is currently too much styling by default. Maybe outlines with the plus on the right are enough as second style? I really liked the G2ish-feeling of this draft:

https://user-images.githubusercontent.com/617986/90201664-5f2c7680-dd90-11ea-9169-f2f9fd8633d8.gif

Everything else could/should be styled by theme or FSE or whatever? :)

**Comment by kellychoffman** [comment]
Technically anything can be done by themes or [insert manual way here] but it would be good if we created some nicely designed "head starts" for people that they could then further tweak as needed. IMO the border is not enough on its own but perhaps pairing that with an alternative layout would be 👍

**Comment by StaggerLeee** [comment]
> ![accordion-flow](https://user-images.githubusercontent.com/617986/89588503-9e8b1e00-d7f8-11ea-974c-75720d1577d3.gif)

I think it is wrong and confusing. Some basic design and CSS should be used for heading and content directly after chosing block type. 
As it is on GIF it strongly reminds of dealoing with simple Paragraph blocks.

**Comment by mapk** [comment]
@mariohamann, I hear you've got some time to begin development on this? Wonderful! Looking back on [your comments above](https://github.com/WordPress/gutenberg/issues/21584#issuecomment-670840252)...

>I decided early on to work with nested blocks

I'm not opposed to this. If you think it would work better, let's try it out. You bring up some positive points there.

>I don't know how useful it is to position the unfolding icon on the left side by default. For example, if someone wants to include icons for orientation in their title, this will be difficult.

I prefer keeping aligned with the expanding icon on the left for now. Allowing users to add their own icon for each accordion element seems quite complex.

>In your design @mapk it looks like the accordion_element in the editor is unfolded when it is selected. This was also the case in my projects at first. It is a common use case that content is written first, but only then is it decided to put it into an accordion. This makes it necessary to access the InnerBlock area of an accordion_element without the accordion_element being currently selected.

If we're building this out as one block with nested blocks, I understand. But I also imagined that if you were dragging content onto an accordion that was collapsed, it would open automatically to allow you to drop it in – similar to folders on an OS.

>I therefore consider your older designs with the minimal border to be a much better default state and should still be minimal enough to make them customizable by themes. From a usability perspective, it will also only then become clear that the complete title section is clickable - which will hopefully be the case? :)

I'd really like to allow a super minimal design variation (no borders). Whether or not it's the default choice could be debatable.

>Concerning the icon, I decided on a plus 

Let's keep the triangle and include the `+` as an option. It looks like the [Figma library](https://www.figma.com/file/e4tLacmlPuZV47l7901FEs/WordPress-Design-Library?node-id=8%3A13) has included a nice triangle in the icons file.

![Screen Shot 2020-09-15 at 2 57 28 PM](https://user-images.githubusercontent.com/617986/93269513-f4da6d80-f763-11ea-9cb0-c8896a51f8c8.png)

I'm looking forward to working with you on this! 👍

**Comment by mariohamann** [comment]
> @mariohamann, I hear you've got some time to begin development on this?

@mapk "A little time" would fit better. ;) I am happy to be involved and I will try to do some work this weekend!

**Comment by mariohamann** [comment]
First of all, I would like to share the concept:

- `accordion` block (being a wrapper) with an `<InnerBlocks />`-component, only `accordion-item`-blocks allowed
- As many `accordion-item`-blocks as needed consisting of e. g. a <RichText/>-component (with no formatting options) and a `<InnerBlocks />`-component

Check out the following GIF to see the concept+example or try it out on [CodePen](https://codepen.io/mariohamann/pen/OJNBGQN). (I am not very happy with the transitions, but I don't think that's important at the moment. :))
![Screen Recording 2020-09-20 at 13 06 01](https://user-images.githubusercontent.com/26542182/93710018-6f7c0200-fb43-11ea-8893-3af1966fcb32.gif)
(By the way: This a great example which would profit very much from https://github.com/WordPress/gutenberg/issues/25133 :))

The blocks I created rely on JavaScript, but at least they should be accessible as the function is based on [W3 accordion example](https://www.w3.org/TR/wai-aria-practices-1.1/examples/accordion/accordion.html). I know, using JS is a big disadvantage. On the other hand there is at least one advantage: The editor can decide, whether `accordion-items` should close as soon as another `accordion-item` is being opened – a very frequently used function and in my opinion a Must-Have.

The JavaScript needs every `accordion-item__button` and `accordion-item__content`  to be connected with their `id` and `aria-controls`/`aria-labelledby`. For this purpose I have  used `useInstanceId` of the `accordion-item`.

**Comment by mariohamann** [comment]
@AlexChariyska DMed me on Slack and we discussed different things  to get rid of JS in frontend. She had the idea of using hidden "checkbox" and "radio inputs". Both would allow accordions to have multiple (checkbox) or only single elements (radio) being open at the same time. At first it felt a bit hacky – but it works surpringly good. Here is a revised frontend prototype: https://codepen.io/mariohamann/pen/QWEErGQ

We are unsure whether this approach is accessible and would love to get some feedback e. g. from @ZebulanStanphill or @enriquesanchez. Of course feedback on the "checkbox/radio" proposal in general would be great, too.

**Comment by alexstine** [comment]
@mariohamann This is not accessible because you set "display: none;" on the input element. Screen readers would only hear the label announced which would not be valid for keyboard accessibility as it would not be focussable by default.

I recommend sticking with the JS front-end approach to toggle aria-expanded attribute. That is how I have always designed accordions and for this solution with checkboxes and radio buttons, I really don't see the big advantage of saving 1-2KB of JS.

Just my thoughts, thanks.

**Comment by mariohamann** [comment]
@alexstine Thank you very much for your fast and great feedback! I've changed the CSS a bit, so the checkbox is selectable. Sorry for this. After testing it again, I realised that the single elements now open and close by using the arrow down and up keys. I think this isn't standard behavior, is it?

Personally I would tend to use JS as well, especially as the JS of W3 seems to be proven accessible. But we couldn't think of any frontend JS in other core blocks and therefore are very unsure, if this should be the way to go. Maybe some other people being closer to the WP core could provide some feedback on this, too.

**Comment by alexstine** [comment]
@mariohamann It does not change with arrow up/down unless the screen reader is in edit/focus mode and attached to the radio buttons. This really just isn't good for UX though as SR users could get really confused if the radio buttons start changing due to the up/down arrow key presses. Most accessible way would be to use the buttons and then toggle aria-expanded.

Thanks.

**Comment by mariohamann** [comment]
Thanks again for your detailed and valuable insights!

So having discussed the following solutions:

1. <summary>/<details>: Not accessible enough and lacks accordion functionality (e. g. closing one item when another one is being opened)
2. checkbox/radio: Not accessible enough and tends to feel hacky
3. JS in Frontend: Provides JS in frontend, which is unusual for core blocks moment, as far as I know.

Are there any other alternatives I didn't mention? If not, we would need a decision, if it's fine to use JS in frontend for this block.

**Comment by alexstine** [comment]
@mariohamann I cannot think of any more possible accessible implementations. Let's go with 3.

Thanks.

**Comment by mtias** [comment]
We can also not add this block in core

**Comment by aristath** [comment]
> I really don't see the big advantage of saving 1-2KB of JS.

With WP powering almost 40% of the web, it makes a **huge** difference for sustainability. 1-2kb on each page-load for 40% of all page-visits on a daily basis adds up to a lot of data... If possible we should strive to save as much as humanly possible.

> We can also not add this block in core

+1

**Comment by mariohamann** [comment]
Added this fourth option in the list above and I think it is definetely a option to go.

@aristath I would expect JS being loaded only on pages with the block inside – not on every page. **Personally** I think that an accordion is one of the most basic elements in today's web and it can't be replaced with any pattern, block combination etc. It should be worth considering to add such an element to the core to prevent users from adding a complete block suite or maybe unaccessible/non-performant accordions only for this basic function.

Therefore: On the one hand I TOTALLY get your point, on the other hand I think it is worth the discussion.

**Comment by pandjarov** [comment]
Accordions are not some extravagant feature but a basic functionality that's part of most pages. If we ever want Gutenberg to be at least on par with the popular page builders out there having one as part of it is a must.

**Comment by alexstine** [comment]
I would also rather see our core implementation of clean JavaScript for accordion than some of the possible plugins that could spin up as a result of this block not being added to core.

**Comment by ZebulanStanphill** [comment]
@mtias
> We can also not add this block in core

Why?

Also, would an accordion block actually require 1-2kB of JS? That seems like a rather large estimate to me.

**Comment by mariohamann** [comment]
> Also, would an accordion block actually require 1-2kB of JS? That seems like a rather large estimate to me.

If we use a minified version of the (accessible) W3 accordion in fact it would be about 1.7kb. But maybe there are better/more performant ones.

**Comment by mtias** [comment]
@ZebulanStanphill it's also an option to consider like the others.

JS on the front is less about the weight but the ongoing support, maintenance, and ramifications it can carry. Core has rarely added front-end JS outside of themes. (Navigation menus in themes was one of the most prominent.) Especially when there is a native and semantic tag to achieve the result with `<details>` and `<summary>` we should consider the implementation thoroughly.

**Comment by mtias** [comment]
Also it might have sounded contrarian, but I do support developing a great accordion block. We could include it in the block directory as a core-supported block if we arrive at a solid implementation instead of bundled, which can also help in the current setup to ensure scripts / styles are bundled separately.

**Comment by ZebulanStanphill** [comment]
@mtias Oh, I misunderstood your previous post.

Unfortunately, because [`<summary>` eats semantics](https://daverupert.com/2019/12/why-details-is-not-an-accordion/) (and there appears to be no way around this, even with JS), I don't think we can use the `<details>` element at all in an Accordion block, if we want to support a11y-navigation-friendly headings.

And, yeah, an officially supported plugin could be a good solution to avoid having the bundle any more JS in core itself.

**Comment by doughamlin** [comment]
@mtias @ZebulanStanphill Are there any other "core-supported blocks" or would this be the first?

**Comment by mariohamann** [comment]
> Particularly when there is a native and semantic tag to achieve some of the result with <details> and <summary> we should consider the implementation thoroughly.

Just in case someone has the same – maybe foolish – idea to add `role` and `level` attributes to the summary tag: It's not HTML compliant and breaks accessibility even more, as far as I understand after some researching and testing.

```html
<details open>
    <summary role="heading" aria-level="2">Item One</summary>
    <div>
         <p>Ut deserunt irure proident fugiat id eu id cupidatat dolore ex.</p>
    </div>
</details>
```

Personally I think if there will be anything in core or core-related/supported which describes itself as being an "Accordion" block, it should provide at least the following two features ([see W3.org](https://www.w3.org/TR/wai-aria-practices-1.1/examples/accordion/accordion.html)):

1. "Allow for multiple accordion sections to be expanded at the same time."
2. "Allow for each toggle to both open and close its section."

For the first point I don't see any possibility for `<details>/<summary>` without JS. If this won't be implemented, in my opinion the block should be named "Toggle" instead of "Accordion" because otherwise it might break WP users expectations.

**Comment by aristath** [comment]
> Just in case someone has the same – maybe foolish – idea to add role and level attributes to the summary tag: lthoug it works (still expandable and being read as e. g. a <h2>), it's not HTML compliant and breaks accessibility even more, as far as I understand after some researching and testing.

Why not just use a proper `<h2>` in the `summary`? AFAIK it's perfectly valid. (apologies if this has already been covered above, but this is a looooong thread and things get lost)

**Comment by mariohamann** [comment]
As @ZebulanStanphill pointed out, it seems to be valid (tested it with [W3 validator](https://validator.w3.org/nu/#textarea)) but not working in terms of a11y: https://daverupert.com/2019/12/why-details-is-not-an-accordion/

[W3 recommends](https://www.w3.org/TR/wai-aria-practices-1.1/#wai-aria-roles-states-and-properties) `<h3><button>Button text</button></h3>`

**Comment by mtias** [comment]
@doughamlin we had discussed it before for things like a _Slider_ block (with inner block container), but this would be the first indeed if we went that direction. As the block directory continues to mature, it could be worthwhile migrating some blocks from core bundling to the directory. We could keep the UX mostly transparent, even showing the blocks in the inserter but they only get installed when you insert and use them.

**Comment by yanceyk** [comment]
Hello. I found this thread about a month ago and forged ahead to find a means to make my own accordion for a client after seeing it in development. 

I was able to do so successfully by using [this tutorial](https://medium.com/skillthrive/build-a-react-accordion-component-from-scratch-using-react-hooks-a71d3d91324b) on Medium. 

I adapted it to pure CSS - no hidden inputs, it relies on the button element. It's accessible to screen readers and works great in the block editor. Here's a pen to the HTML/CSS version.
[https://codepen.io/kyancey/full/KKMQazw](https://codepen.io/kyancey/full/KKMQazw)

Hope it helps.

**Comment by mariohamann** [comment]
Thank you very much @yanceyk for your input! It's very interesting to see those different approaches!
- In Safari your Codepen has some problems – the accordions stayed only open while clicking. After releasing the mouse they closed.
- Still I'm not sure if this accessible, as it uses "focus" in addition to "active" and that would open accordions while tabbing through the content.
- Furthermore I don't think that it's possible to wrap the buttons with a heading tag (e. g. `<h3><button>Lorem Ipsum</button>/h3>`) as the `:active` couldn't be used anymore.

This leads me to the point: There are so many different approaches to create accordions and often they feel robust or accessible – but often they maybe aren't. → A Core supported, accessible accordion is urgently needed.

**Comment by yanceyk** [comment]
Thanks for your feedback @mariohamann, I hadn't tested in Safari. But it was the intent to have them remain open while clicked. Yeah, it isn't possible to wrap a button with `<h3></h3>`. Just thought I'd give you guys some other approaches to consider. I agree with you 100%. I hope you guys are able to come up with something satisfying.

**Comment by LukaszJaro** [comment]
Is there also consideration for the accordion block to allow transforms? i.e, groups/columns transform to accordion?

Now I realized I should of wrote "Accordion" as a essential plugin in the WordPress 2020 survey..

**Comment by paaljoachim** [comment]
How the heck do we move forward with this issue? 
As it would be very helpful to have an Accordion block in core Gutenberg.
What is a method that we can actually agree on? That works with no js and is also accessible.

**Comment by LukaszJaro** [comment]
I was hoping gutenberg block editor to achieve minimal, basic, and useful features compared to other modern page builders.

The argument that it would add JS bloat doesn't hold well. User will just find another plugin or block which may contain 20 other unneeded blocks or bloated JS.

Even an officially supported block will not be found by many and will resort to other bloated plugins. But this seems to be the only middle-ground since no core blocks should have any JS to maintain, from my understanding that is. Also will reduce maintenance on the core team. Maybe through the block directory?

I think some facts have to be solidified here as well so we can move on:

- We can't use details/summary, or checkbox, none of that is accessible enough.
- We can't use pure CSS solution, not accessible, needs the aria expanded value for example.
- Valid example with JS from W3 https://www.w3.org/TR/wai-aria-practices/examples/accordion/accordion.html

**Comment by mtias** [comment]
To clarify, it's alright to have blocks that require JS to run on the front-end in core, but it demands a more in depth consideration and more certainty that it's the right path (including longer term web semantics) since maintenance, back-compatibility, and plugin/theme compatibility costs are higher.

> Even an officially supported block will not be found by many and will resort to other bloated plugins.

An officially supported block through the block directory would show up in the inserter when you search for it as the canonical result, I'd imagine somewhat similar to how the core importers work as separate plugins. There are other candidates in the current set of blocks that could be moved there once the directory is ready for it, for example.

**Comment by LukaszJaro** [comment]
Thanks @mtias for clarifying!

As for the web semantics W3(the people who set web standards) have already figured this out with examples https://www.w3.org/TR/wai-aria-practices-1.1/#accordion

An officially supported block is starting to sound better...

**Comment by paaljoachim** [comment]
@mkaz brought up this PR during the Core Editor Meeting:
https://github.com/WordPress/gutenberg/pull/29537
From how I understood it the above PR will make the Accordion block with JS possible in Gutenberg.

Link to the discussion:
https://wordpress.slack.com/archives/C02QB2JS7/p1615386204074800

**Comment by paaljoachim** [comment]
A nice resource to take a closer look at:
https://semantic-ui.com/modules/accordion.html

**Comment by kraftner** [comment]
@paaljoachim Looked at it for 2 seconds and already did see:

- Breaks down without JS (no progressive enhancement)
- Can't be navigated with the keyboard
- Generally it's all `<div>`s and not even the hint of care for accessibility

**Comment by ciampo** [comment]
Should we consider adding an Accordion component to the `@wordpress/components` package, which then can be reused across the repository (i.e in the Accordion block, the Navigation block #29747...)?

**Comment by mtias** [comment]
@ciampo no, our UI components are React based intended for admin UI, an accordion block would be something implemented entirely differently and optimized for the front-end.

**Comment by ciampo** [comment]
> an accordion block would be something implemented entirely differently and optimized for the front-end

Makes sense, thank you for clarifying.

**Comment by LukaszJaro** [comment]
I'm seeing alot of WordPress plugins bundle accordion blocks, and 90% of them are using the "Details" element unknowingly/knowingly creating accessibility issues so would definitely be nice if WP released a good example of an accessible accordion.

How important is progressive enhancement with accordions? Bootstrap CSS library, known for accessibility, has a  collapse(accordion) component. In their implementation, if you disable JS in browser, the accordion functionality breaks, meaning you can't expand or collapse.

`While WCAG 1.0 from 1999 required that pages be functional and accessible with scripting disabled, WCAG 2 and all other modern guidelines allow you to require JavaScript, but the scripted content or interactions must be compliant with the guidelines`

Even the w3.org accessible accordion design does not work when JS is disabled: https://www.w3.org/TR/wai-aria-practices-1.1/examples/accordion/accordion.html

And then there's WordPress's accordion panels in the admin which also break with JS disabled.

Is this considered a blocker for WordPress's implementation of a accordion block? Meaning there should be some sort of fallback if JS is disabled.

Here's a good resource on this:
https://meiert.com/en/blog/must-work-without-javascript/

If it is necessary for progressive enhancement, then one the strongest and most accessible accordions with progressive enhancement is detailed here https://inclusive-components.design/collapsible-sections/  and here's a codepen https://codepen.io/heydon/pen/ZXdbxj/

It basically uses web components to solve the issue.

**Comment by westonruter** [comment]
There is a standard HTML feature that can be used to implement accordions today: the [`details` element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details) ([Can I use](https://caniuse.com/details)). Not only does it work with JavaScript disabled, but with https://github.com/whatwg/html/pull/6466 browsers will automatically expand `details` elements when doing find-in-page if there is text that matches inside, or if there the page is linking to an anchor inside the `details`. So `details` is both accessible to when JS is turned off _and_ it's accessible for users to easily find text inside of them. The only progressive enhancement for a `details`-powered accordion would be whether the author wants only one accordion to be open at a time. That could be handled with a minimal amount of scripting.

**Comment by ZebulanStanphill** [comment]
@westonruter But there's a catch: you can't put a heading in the `<summary>`.

https://daverupert.com/2019/12/why-details-is-not-an-accordion/

EDIT: It's worth noting that there is an effort to design and propose a real accordion element for HTML, but don't expect it to be in browsers any time soon. https://daverupert.com/2021/10/native-html-tabs/

**Comment by alexstine** [comment]
Quick question. Is there a reason we're working this hard to avoid a few lines of JavaScript on the front-end? I am not at all advocating to make JS everywhere on the front-end but there are plenty of stylesheets at play. I am not seeing where the JavaScript is creating such a big implementation issue.

```
<button type="button" aria-expanded="false" id="accordion1" class="button"><span class="icon-class"></span><span class="screen-reader-text">Some Label</span></button>
<div style="display:none;" id="accordion1-content">
<p>Some content</p>
</div>
```

Could include a custom JS file that gets minified.

```
const accordion = document.getElementById( 'accordion1' );
accordion.addEventListener( 'click', function() {
if ( accordion.classList.contains( 'open' ) ) {
// Remove class and toggle aria-expanded to false.
} else {
// Add class and toggle aria-expanded to true.
}
});
```

This is just a basic example, but I can't imagine this code minified would even make a difference. Besides, it's about good accessibility and functionality. Win win to me.

Thanks.

**Comment by mtias** [comment]
Part of the problem has been that there's a lot being put into the "accordion" bucket and a single block trying to do it all will probably be ill-advised at the moment because people have different expectations on what it should be (a semantic element, a UI component, auto-collapse for siblings, etc).

@alexstine that's a good question! Your example is fairly innocuous on the surface, but it starts to be a bit harder to handle if you have 100 accordions on a page (if using an id as shown above). There are also issues with attaching the event listeners in cases like async loading posts (sites with infinite scroll, etc). This doesn't mean it's a huge blocker but for a block provided by core it would need to be handled somehow and maintained in a way that doesn't cause issues in the long term. People would want to extend it for auto-collapsing creating further couplings with markup and even handlers. The problem is not so much that it's hard to achieve, but that it's not clear — given this whole thread — what exactly everyone has in mind as being a sufficient experience.

In that sense, I still think an accordion, even if maintained by core, is probably best started as a single block in the block directory while these things get cleared. Bundling in core, if deemed necessary, would be easy to do at any point; unbundling wouldn't.

As an aside, I also think `details` with `summary` is something we should probably have in addition to any accordion as a standalone block without heading support under a different name. It seems valuable and useful on its own as a semantic set of elements, it's native html already, works with searches, etc, even if not used as a full fledged sectioning accordion.

**Comment by ZebulanStanphill** [comment]
There was already a PR for a `<details>` block (#23940), but it was closed due to the "overall excitement around the accordion block", heh. 🥴

There was also an issue with handling the spacebar key in the `<summary>` in the editor, but I think that's been resolved now.

If desired, I could open a new PR based on the prior one.

**Comment by mtias** [comment]
We might need a better name, but I think we can reopen :)

**Comment by kimclowa8c** [comment]
Hi everyone. 
I'm an account manager on Team 51; and would love to reopen this and try get an accordion into core. 

From reading above, am I correct in saying we're stuck here: 
- The details block functions nicely; but itsn't accessible, therefore we can't proceed with it? 
- We ideally would like a block that doesn't rely too heavily on js, as this could get complicated if we were to have multiple accordions on a page  

We have built a few custom accordion blocks for sites we've built - this one you can see on the front end below: 
- See the FAQ section at the bottom of: https://dayoneapp.com/pricing/ 

We're about to move Day One App over to WPCOM Simple (it's currently on Pressable), so that accordion will be included as custom theme in the code, perhaps that can lay a foundation for a core block? 

And then others I can see in our custom block library, but not on the front end:
- https://github.com/a8cteam51/resumido/blob/trunk/mu-plugins/stag-blocks/src/blocks/accordion/index.js 
- https://github.com/a8cteam51/swami-select/blob/trunk/mu-plugins/swami-select-blocks/src/blocks/hidden-content/index.js
- https://github.com/a8cteam51/language-please/blob/trunk/plugins/blocks/accordion-block/index.js 

Let me know if you need any permissions granted to access our team repo.

**Comment by jordesign** [comment]
@ZebulanStanphill Just coming across this - and wondering what would be involved in reponening that PR for discussion?

Would be great to get it going again.

**Comment by ZebulanStanphill** [comment]
@jordesign Recently, the special handling for pressing the space key in the `<summary>` element [was removed](https://github.com/WordPress/gutenberg/pull/41977).

In order to have a `<details>` block, you'd have to use different markup in the editor versus the front-end. The trouble is that you want the fake `<details>` to look like the real one... but how do you pull that off when a `<details>` looks different depending on your browser? And then there's the challenge of ensuring that any styles applied to a `<details>` on the front-end are also applied to the fake one.

For these reasons, I got stumped while trying to make a new PR to add a `<details>` block.

A proper accordion block (which due to previously stated reasons, wouldn't use a `<details>`), probably wouldn't have most of these markup/styling mismatch issues. However, it has its own challenges, namely how to handle the custom JS that would be loaded on the front-end. Ironically, though, it's starting to sound a lot easier than a `<details>` block.

**Comment by Jabe64** [comment]
@carolinan Any update on accordion block now that details/summary block seems to be done?

I don't know if this could help but there is a nice video about doing accordion with great accessibility with less JS and more CSS:
https://www.youtube.com/watch?v=WJERnXiFFug

The final code is here : https://github.com/kevin-powell/accordion

There is also another shorter video about using CSS grid-template-rows property instead of JS for transitions:
https://www.youtube.com/watch?v=B_n4YONte5A

Final code : https://codepen.io/kevinpowell/pen/NWOgVga

**Comment by westonruter** [comment]
Just wanted to note that Chromium just published an [intent to prototype the accordion pattern using `<details>` elements](https://groups.google.com/a/chromium.org/g/blink-dev/c/Q1OX6ZA_aaE/m/ALwkAOfHAwAJ). This could end up being a way that the accordion becomes part of the web platform, and as such, might be worth waiting to see what comes of it in regards to creating an Accordion block.

**Comment by Jabe64** [comment]
Thanks for the info @westonruter 
I have just realized that Details Block isn't finished. There are more refinements coming in #49808

**Comment by carolinan** [comment]
@Jabe64 thanks for asking. The details block proposal was not accepted, and where either block goes from here will be up to others.

**Comment by Jabe64** [comment]
> @Jabe64 thanks for asking. The details block proposal was not accepted, and where either block goes from here will be up to others.
> 

You're welcome! Thanks for the update and fast answer. Maybe what @westonruter said would be of interest of the details block conversation.

**Comment by mrwweb** [comment]
A la @westonruter's point about watching changes to Chrome's `details`/`summary` support, I also wonder about how the progress of the [Interactivity API](https://make.wordpress.org/core/2023/03/30/proposal-the-interactivity-api-a-better-developer-experience-in-building-interactive-blocks/) could impact this block. In many ways, an accordion is _the_ best example of a similar, interactive block that is a perfect use case for the API. If it looks like the API is going to successfully make it into core, it could be really cool to pair it with the release of a core block that puts it to use.

_Further_, even with the potential changes to `details`/`summary`, I'm not sold that it's the right tool for this block given that you can't use headings to identify accordion sections with `details`/`summary`.
[
Scott O'Hara has an _excellent_ overview of `details`/`summary`](https://www.scottohara.me/blog/2022/09/12/details-summary.html) and concludes with the following:

> It really just all comes down to what sort of experience you want to provide to people, and how OK you might be knowing that the native elements are going to be exposed differently, have different gaps but also different features to them. Or, you can roll your own disclosure widgets and provide a consistent experience that you are fully in charge of… but it may not be as feature rich as what the browsers could provide, and your disclosure widgets need to degrade gracefully if viewed in a browser’s reading mode.
> 
> The decision of what you do is up to you.

Using those elements for the core block is probably defensible, but the rationale for doing that instead of a custom implementation with minimal JavaScript (as @alexstine mentioned earlier in this thread) should be clearly documented and justified.

**Comment by cbravobernal** [comment]
I will be happy to try an experiment of an Accordion approach with the [Interactivity API](https://make.wordpress.org/core/2023/03/30/proposal-the-interactivity-api-a-better-developer-experience-in-building-interactive-blocks/). I will keep you posted in this issue.

**Comment by jordesign** [comment]
@mapk @paaljoachim  Just checking in here to see if the new Details block - covers this issue sufficiently

**Comment by paaljoachim** [comment]
Hey @jordesign 
I have no idea....

**Comment by jordesign** [comment]
I'm inclined to say the the new Details Block handles this well enough to close this issue - any enhancements/suggestions to extend the functionality could be added as new Enhancement reauests

**Comment by joedolson** [comment]
The details block is really not a substitute for an accordion, and should not be used in that manner. It's expected to show and hide only one panel, and doesn't operate as a group. I think we should still add an accordion block.

**Comment by amberhinds** [comment]
I was just going to write the same thing RE the details block. Here's a good [example of how the accordion block should be coded for accessibility](https://designsystem.digital.gov/components/accordion/).

**Comment by webd-uk** [comment]
> The details block is really not a substitute for an accordion, and should not be used in that manner. It's expected to show and hide only one panel, and doesn't operate as a group. I think we should still add an accordion block.

I'd like to highlight that the new Twenty Twenty-Four theme uses multiple Details Blocks to provide the functionality of an accordion on the "FAQ" panel of the home page.

The _only_ thing missing from the Details Blocks is a setting "Close other Details Blocks when opened" which would negate the requirement for an Accordion Block.

So ... if the Details Block shouldn't be used to create an accordion then really the new theme shouldn't be promoting it as such surely?

**Comment by LukaszJaro** [comment]
@webd-uk I would a ticket for that.

@alexstine would this be considered accessibility issue with the way the 2024 theme is using the details block as accordion in FAQ section? Also what uses cases should the details element be used?

**Comment by webd-uk** [comment]
@LukaszJaro I did with #55483 ... shall I re-open it? Or do you mean raise a ticket about the Twenty Twenty-Four theme use of the Details Block? Because if you do I'd rather not as having played with the Details Block I think it's perfect for an accordion if it can just have the option added to auto-close other Detail Blocks.

**Comment by joedolson** [comment]
The 2024 theme is using a series of details blocks in a group, but I don't see any way that they're describing it as an accordion. The fact that a group of details are used in sequence does not make it an accordion, and I don't see anything that promotes it as an accordion. If you can point to some specific text in which the FAQ pattern is described as or promoted as an accordion, I'm happy to push back on that, but while I wouldn't *use* this pattern, because it has woefully insufficient semantics for what it represents, it's not described as an accordion, as far as I can tell.

**Comment by joedolson** [comment]
@webd-uk Re: "The only thing missing from the Details Blocks is a setting "Close other Details Blocks when opened" which would negate the requirement for an Accordion Block."

This is not true. The thing that's missing from the details block is heading semantics. An accordion should have headings that indicate each item within the accordion. Inside those headings, there should be buttons that trigger the accordion behavior.

The details block has a `summary` element, which has the role of `button` natively. The button role overrides all internal semantics, so no matter what content is inside the `summary` element, there are no headings present. That is the crucial reason that `details` is not a substitute for an accordion. As soon as you have enough content that headings are an important element for navigating content, `details` ceases to be sufficient.

**Comment by webd-uk** [comment]
@joedolson Ah, OK. So adding a "Close other Details Blocks when opened" setting to the Details block would not make it behave like an accordion because without the heading semantics it cannot be an accordion. As such, do you think that #55483 is valid and should I should re-open it?

**Comment by joedolson** [comment]
No, I think that #55483 is not a valid request. There shouldn't be any relationship between different `details` elements on a given page; they should each be treated as an independent element. A change like that would just increase the impression that the details block is equivalent to an accordion.

**Comment by webd-uk** [comment]
OK ... in that case I'll be looking forward to the forthcoming Accordion Block!

Thank you.

**Comment by ocean90** [comment]
The `details` element can now have a `name` attribute to make it [an "exclusive accordion"](https://developer.chrome.com/docs/css-ui/exclusive-accordion?hl=en#the_exclusive_accordion).

**Comment by joedolson** [comment]
That makes the behavior more like an accordion, but it's still significantly lacking in semantics. It would function in an accordion style, but it would be a very poor accordion. In principle, I'd support adding the attribute on the grounds that WordPress should support all standardized HTML. However, I would still stand on the opinion that it is not an accordion, and we should provide an accordion that does meet accessibility standards.

**Comment by westonruter** [comment]
@joedolson What are the lacking semantics? Are they lacking in the [spec](https://html.spec.whatwg.org/multipage/interactive-elements.html#the-details-element) or not implemented in browsers or what?

**Comment by amberhinds** [comment]
@westonruter take a look at this [example of how the accordion block should be coded for accessibility](https://designsystem.digital.gov/components/accordion/) that I shared in September.

**Comment by joedolson** [comment]
The most specific gap is that an accordion should be a list of heading elements that contains buttons to expand and collapse the content.

The semantics of the details/summary elements, however, are designed so that you can only place a heading inside the `summary` element; you can't wrap the `summary` in a heading without breaking the control. This reverses the semantic relationship, and removes the heading semantics.

A heading with a contained button is a heading that has text equivalent to the button text plus a button that has the interactivity. 

A summary elements with a contained heading, however, is an interactive element that does not contain any heading semantics.

As an isolated element, a single summary/details combination is a great, easy to implement disclosure widget. But as a group of related elements, it's lacking key navigation and structure elements. 

I don't off-hand know whether this should be primarily blamed on browser vendors or on AT vendors. The spec doesn't provide clear guidance about how headings should be handled within a `summary` element, and as of the last testing publically available, JAWS offered no support for headings within `summary` on any browser, which eliminates a very large percentage of users.

[That test dates to May 2022](https://a11ysupport.io/tests/tech__html__details-summary), though, so it may be due a revisit.

**Comment by LukaszJaro** [comment]
@c4rl0sbr4v0 how did the experiment go with the interactivity api?

> I will be happy to try an experiment of an Accordion approach with the [Interactivity API](https://make.wordpress.org/core/2023/03/30/proposal-the-interactivity-api-a-better-developer-experience-in-building-interactive-blocks/). I will keep you posted in this issue.

**Comment by cbravobernal** [comment]
I am now working on making[ the Interactivity API public in WordPress 6.5.](https://github.com/WordPress/gutenberg/issues/56803)

Until then, I won't have enough time 😅 . But after that, it would be a great example for this new feature!

**Comment by Nyiriland** [comment]
🖐️ Gey gang! I work in an agency and we use accordions quite a bit on our projects. It would be amazing if we could use something from core instead of resorting to a plugin, or building out custom blocks (pretty sure we've done both in the past). Accordions are a very common layout feature on modern sites, for things like feature lists, or FAQs.

The reason Details doesn't fulfill the same purpose for me is two-fold:
- It looks like a bullet point (and I thought it was until someone pointed out it wasn't)
- A user would need CSS knowledge to be able to update it to a more "modern" styled accordion; right now it feels pretty dated as-is (and I'm pretty sure it's just using the default html styles).

**Comment by jarekmorawski** [comment]
Love the discussion so far! While working on some patterns for WooCommerce, I found the Details block limiting and would love to update it with more features or build a dedicated accordion block. I put together a design proposal that solves most of the issues I ran into and most of what @Nyiriland described earlier).

![Details](https://github.com/WordPress/gutenberg/assets/79307566/ff8e1caf-475d-431a-8b1e-a04d26b97b05)

Firstly, the block would consist of two parts: **a trigger and a hidden content area**. Both could be anything: from buttons to images to text to tables. Clicking on the trigger block would open the accordion.

![Default](https://github.com/WordPress/gutenberg/assets/79307566/3f9b245d-fcd2-49c3-83ac-0259e0018596)

For instance, we'd let users add a button that will open the accordion.

![Button](https://github.com/WordPress/gutenberg/assets/79307566/bbc09acf-adc0-44b0-8bc2-1ae3d3ad3910)

Going further, if we update the button to support icons and offer some extra layout flexibility (like "Stretch" as a justification setting that will spread the button's inner blocks even across its width), we'd make it easy to build modern-looking accordions with chevrons, -/+ buttons, text labels, etc.

![Filter](https://github.com/WordPress/gutenberg/assets/79307566/7d93b92f-7a69-4b8f-9668-57624fc51965)

As for the content area, it could contain **any number of blocks**. Here's how the block structure would look if we were to build an accordion for a price filter block in a product catalog page template in Woo.

![Structure](https://github.com/WordPress/gutenberg/assets/79307566/55817ccb-795e-4fe4-bbaf-bd5dfd5fa61c)

We'd use block states (related to #57719, #38277) to let users customize the appearance of a collapsed and expanded accordion. This would enable designers and builders to use different text labels or icons for each block state, e.g., `-` when the accordion is open, and `+` when it's closed.

![States](https://github.com/WordPress/gutenberg/assets/79307566/db77dbe2-ecf7-4e11-a482-fa9546d9c44c)

We'd also let users **decide if the accordion is open by default and whether other accordion blocks on the page should collapse automatically** when this one's open (this is a pretty common behavior on the web).

I'm a bit torn about this part because it can cause issues when enabled on multiple blocks, but then building an accordion group parent block just for this setting would add unnecessary complexity.

![Settings](https://github.com/WordPress/gutenberg/assets/79307566/65ef4196-58f2-438f-a030-3565436c235d)

We would let users specify the block's **width, justification, and orientation** to enable more creative, non-standard layouts. For instance, they'd design an accordion where the trigger is on the left or right, which is a common sight in e-commerce.

![Layout](https://github.com/WordPress/gutenberg/assets/79307566/1aeb2ad0-c09a-4797-89d9-ebdb7b7c9376)

**Comment by portrayaloflife** [comment]
What needs to happen for this to be added to core? This would be such a useful feature.

**Comment by vk17-starlord** [comment]
I've been following this thread and the interest around an Accordion block for quite some time. I see that it's a highly requested feature, and the conversation around accessibility, semantics, and native elements like `<details>` has been super insightful.

I'd love to contribute and was wondering:

- Is there a current owner actively working on this?
- Would it be helpful if I took a stab at creating a prototype for this block?
- Should the prototype aim to be a minimal MVP (possibly using native `<details>` for accessibility) or lean toward something more fully featured with customization options?
- Are there any design or UX directions the team would prefer to be followed (maybe something that aligns with Interactivity API plans)?
- Is there a Figma or issue with design specs I should reference?

If it's okay, I’d be happy to start experimenting and contribute a PR for early feedback.
Let me know what you think! 🙌  

Recently worked on prototype - [Back to top component](https://www.loom.com/share/508f6232ff2d48488978a95481063322?sid=091509b4-686c-48c0-b480-c95325dbfd17) for https://github.com/WordPress/gutenberg/pull/69911

— Vinit

**Comment by webd-uk** [comment]
@vk17-starlord We've been waiting for someone to develop this too. In the meantime, we use several Details blocks and some custom JavaScript to link them together. Scroll down to "> Recent Events" to see a working example on this website https://www.stephsices.co.uk

I realise this isn't what the purists consider to be an accordion but nonetheless our clients are happy :)

**Comment by vk17-starlord** [comment]
Hi @Mamaduka @joedolson – apologies for tagging you again, but I wanted to kindly ask if this issue is still relevant and worth working on. I'm still new here and really appreciate your guidance. I tagged you both since you're core contributors and have a deep understanding of the project. Thanks in advance!

**Comment by amberhinds** [comment]
This is still relevant, and it would be great to see this built. @vk17-starlord, please read through all of the comments above. The details element should not be used for grouped accordions.  There needs to be a setting for the block that allows people to set the heading level for the accordion title. 

Here's an example of expected HTML output and functionality: [How to Code Accessible Accordions](https://equalizedigital.com/how-to-code-accessible-accordions/)

**Comment by westonruter** [comment]
> The details element should not be used for grouped accordions.

Is this still the case, considering that the accordion use case is now [explicitly supported](https://developer.mozilla.org/en-US/blog/html-details-exclusive-accordions/) in the `details` element via the `name` attribute? See also https://github.com/WordPress/gutenberg/issues/56969 and https://github.com/WordPress/gutenberg/pull/56971.

**Comment by joedolson** [comment]
It probably needs some more exhaustive testing, but it is now possible to use a series of `details` elements to create an accordion. It's not as simple as just using them and adding the `name` attribute, however; it still requires a context wrapper with a name (e.g. a `section` element. There are a number of preferred behaviors that an accordion should support that aren't supported out of the box; like scrollable panels. That said, these are all issues that also need to be dealt with using JS in an accordion that *doesn't* use details/summary, so that's not a blocker. 

But the headings issue is largely obsolete; headings inside `summary` elements are now supported in most contexts. (Possibly all major contexts; but I didn't test it exhaustively enough to say that for sure.) But they are reported as headings, and are navigable using screen reader heading commands. 

As the Details block currently works, however, it's not viable for an accordion; it needs support for headings inside the summary.

**Comment by kimclow** [comment]
In case this is helpful, there is an open PR to add an accordion to Core here: https://github.com/WordPress/gutenberg/pull/64119 

It's pretty close to being ready to deploy, just a handful of checks left.

**Comment by alexandrebuffet** [comment]
Hello there! Great work! It’s really exciting to see this block in the inserter! 🎉

I know this is already merged, but I’d like to comment on the block’s name.

The block was renamed from "Accordion Group" to "Accordions" (plural), and each element is an "Accordion Item".

While "Accordions" aligns with other plural-named blocks like Columns and Buttons (as @mikachan mentioned in PR #64119), I think it still can be confusing. 

This block represents a single accordion component, a group of collapsible items. So "Accordion" (singular) would be more accurate IMO.

Above all, using singular name for the block would align with the W3C WAI-ARIA Accordion pattern, and make it immediately clear to users familiar with the term. See: https://www.w3.org/WAI/ARIA/apg/patterns/accordion/.

This naming convention is also used in most UI libraries (Material UI, Radix UI, Bootstrap, etc.).

What do you think?

**Comment by mikachan** [comment]
Thanks for the feedback, @alexandrebuffet.

The Accordion Item inside the Accordions block is currently called "Accordion", so I think we would need to rename this to "Accordion Item" (or something else) in the inserter, if we were to rename the group to "Accordion":

<img width="349" height="289" alt="Image" src="https://github.com/user-attachments/assets/99e1537e-40ed-42a0-95c9-1b195f8bc475" />

I agree that this would align with how W3C describes an Accordion, so I'll open a PR for this change and ping the folks who input into the name in the original PR. Thanks!

Edit: PR opened here: https://github.com/WordPress/gutenberg/pull/71222

**Comment by t-hamano** [comment]
My biggest concern is how users will understand the difference between the Details block and the Accordion block if the Accordion block is stabilized and shipped to core.

A consumer who is knowledgeable about HTML and accessibility will likely understand the purpose of each block and be able to use each block "correctly". However, average users may wonder why there are two similar blocks. Is there an efficient way to communicate the difference between the two blocks to users?

**Comment by joedolson** [comment]
One thing that would help would be if the details block did not register on a search for 'accordion'.  I argued against this in https://github.com/WordPress/gutenberg/pull/53501, and with the addition of an actual accordion block, it's even more critical.

**Comment by itsViney** [comment]
Great to see this block make it to core after all this time. Good work to all involved.

I don't know if this is within the scope of Gutenberg or if it should be picked up by plugins like Yoast, but given one of the most common use cases for an accordion block is an FAQ, I wonder if it would be a good idea to add a toggle for 'Enable FAQ schema markup'?

**Comment by kimclow** [comment]
Nice idea @itsViney - and also perhaps a pattern we can include in Core.

**Comment by itsViney** [comment]
> Great to see this block make it to core after all this time. Good work to all involved.
> 
> I don't know if this is within the scope of Gutenberg or if it should be picked up by plugins like Yoast, but given one of the most common use cases for an accordion block is an FAQ, I wonder if it would be a good idea to add a toggle for 'Enable FAQ schema markup'?

If anyone else would find this handy, I knocked together a simple plugin to enable FAQ schema markup on the Accordion block: https://github.com/itsViney/accordion-faq-schema-toggle.

<img width="1232" height="496" alt="Screenshot showing the Output FAQ Schema toggle in the Accordion block's settings panel." src="https://github.com/user-attachments/assets/2d209a8e-75b0-4cec-becb-b40b17ff4a66" />

I'm not confident enough to contribute directly to Gutenberg yet but if anyone fancies incorporating it, feel free to use this (very basic) code as a starting point.

**Comment by jupa8712** [comment]
Hi, I just tested the accordion block, and I was surprised to find it's made with divs. I thought it would be a native element (details) and that details would be used. Something like this example I have here. 

<div>
  <details name="tech-specs">
    <summary>Title 1</summary>
    <p>Text 1</p>
  </details>
  <details name="tech-specs">
    <summary>Title 2</summary>
    <p>Text 2</p>
  </details>
  <details name="tech-specs">
    <summary>Title 3</summary>
    <p>Text 3</p>
  </details>
</div>

Here an example with css animation: https://codepen.io/jupa8712/pen/wBMmKOP

Is there any compelling reason? Thank you very much.

**Comment by Mamaduka** [comment]
There's also a Details block; semantically, it serves a different purpose.

**Comment by mikachan** [comment]
> I just tested the accordion block, and I was surprised to find it's made with divs.

+1 to @Mamaduka's comment, and we also followed the HTML semantics and practices presented in these examples: [WCAG](https://www.w3.org/WAI/ARIA/apg/patterns/accordion/examples/accordion/) and [USWDS](https://designsystem.digital.gov/components/accordion/).
